File: renderer/src/resume.rs
use serde::Deserialize;

/// A JSON Resume document ([`jsonresume/resume-schema`]).
///
/// Unknown properties are accepted and ignored by the HTML emitter.
#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Resume {
    #[serde(rename = "$schema", default)]
    pub schema: Option<String>,
    #[serde(default)]
    pub basics: Option<Basics>,
    #[serde(default)]
    pub work: Option<Vec<Work>>,
    #[serde(default)]
    pub volunteer: Option<Vec<Volunteer>>,
    #[serde(default)]
    pub education: Option<Vec<Education>>,
    #[serde(default)]
    pub awards: Option<Vec<Award>>,
    #[serde(default)]
    pub certificates: Option<Vec<Certificate>>,
    #[serde(default)]
    pub publications: Option<Vec<Publication>>,
    #[serde(default)]
    pub skills: Option<Vec<Skill>>,
    #[serde(default)]
    pub languages: Option<Vec<Language>>,
    #[serde(default)]
    pub interests: Option<Vec<Interest>>,
    #[serde(default)]
    pub references: Option<Vec<Reference>>,
    #[serde(default)]
    pub projects: Option<Vec<Project>>,
    #[serde(default)]
    pub meta: Option<Meta>,
}

impl Resume {
    /// Deserialize a JSON Resume document.
    ///
    /// # Errors
    ///
    /// Returns the `serde_json::Error` when `json` is malformed or a field
    /// has a type these structs do not accept (unknown fields are ignored).
    pub fn from_json(json: &str) -> Result<Self, serde_json::Error> {
        serde_json::from_str(json)
    }
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Basics {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub label: Option<String>,
    #[serde(default)]
    pub image: Option<String>,
    #[serde(default)]
    pub email: Option<String>,
    #[serde(default)]
    pub phone: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub location: Option<Location>,
    #[serde(default)]
    pub profiles: Option<Vec<Profile>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Location {
    #[serde(default)]
    pub address: Option<String>,
    #[serde(default)]
    pub postal_code: Option<String>,
    #[serde(default)]
    pub city: Option<String>,
    #[serde(default)]
    pub country_code: Option<String>,
    #[serde(default)]
    pub region: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Profile {
    #[serde(default)]
    pub network: Option<String>,
    #[serde(default)]
    pub username: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Work {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub location: Option<String>,
    /// Company tagline. Stored on the Resume; not rendered in HTML contract 1.0.
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub position: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub start_date: Option<String>,
    #[serde(default)]
    pub end_date: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub highlights: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Volunteer {
    #[serde(default)]
    pub organization: Option<String>,
    #[serde(default)]
    pub position: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub start_date: Option<String>,
    #[serde(default)]
    pub end_date: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
    #[serde(default)]
    pub highlights: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Education {
    #[serde(default)]
    pub institution: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub area: Option<String>,
    #[serde(default)]
    pub study_type: Option<String>,
    #[serde(default)]
    pub start_date: Option<String>,
    #[serde(default)]
    pub end_date: Option<String>,
    /// `"3.7"`, `3.7`, or `4` in the wild; always text here (`"3.7"`, `"4"`).
    #[serde(default, deserialize_with = "string_or_number")]
    pub score: Option<String>,
    #[serde(default)]
    pub courses: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Award {
    #[serde(default)]
    pub title: Option<String>,
    #[serde(default)]
    pub date: Option<String>,
    #[serde(default)]
    pub awarder: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Certificate {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub date: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub issuer: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Publication {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub publisher: Option<String>,
    #[serde(default)]
    pub release_date: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub summary: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Skill {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub level: Option<String>,
    #[serde(default)]
    pub keywords: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Language {
    #[serde(default)]
    pub language: Option<String>,
    #[serde(default)]
    pub fluency: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Interest {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub keywords: Option<Vec<String>>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Reference {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub reference: Option<String>,
}

#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Project {
    #[serde(default)]
    pub name: Option<String>,
    #[serde(default)]
    pub description: Option<String>,
    #[serde(default)]
    pub highlights: Option<Vec<String>>,
    #[serde(default)]
    pub keywords: Option<Vec<String>>,
    #[serde(default)]
    pub start_date: Option<String>,
    #[serde(default)]
    pub end_date: Option<String>,
    #[serde(default)]
    pub url: Option<String>,
    #[serde(default)]
    pub roles: Option<Vec<String>>,
    #[serde(default)]
    pub entity: Option<String>,
    #[serde(default)]
    pub r#type: Option<String>,
}

/// A JSON string or number, as an untagged shape for `string_or_number`.
#[derive(Deserialize)]
#[serde(untagged)]
enum StringOrNumber {
    String(String),
    Number(serde_json::Number),
}

impl From<StringOrNumber> for String {
    fn from(value: StringOrNumber) -> Self {
        match value {
            StringOrNumber::String(text) => text,
            StringOrNumber::Number(number) => number.to_string(),
        }
    }
}

/// Accept a JSON string or number where the schema says string. Integers
/// print without a decimal point (`4`); floats print as written (`3.7`).
fn string_or_number<'de, D>(deserializer: D) -> Result<Option<String>, D::Error>
where
    D: serde::Deserializer<'de>,
{
    Option::<StringOrNumber>::deserialize(deserializer).map(|value| value.map(String::from))
}

/// JSON Resume `meta`. Never rendered. Extra keys (`x-schema-resume`, …) stay here.
#[derive(Debug, Clone, Default, PartialEq, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Meta {
    #[serde(default)]
    pub canonical: Option<String>,
    #[serde(default)]
    pub version: Option<String>,
    #[serde(default)]
    pub last_modified: Option<String>,
    #[serde(flatten)]
    pub extra: serde_json::Map<String, serde_json::Value>,
}

#[cfg(test)]
mod tests {
    use super::*;

    fn score_of(json: &str) -> Option<String> {
        let education: Education = serde_json::from_str(json).expect("valid education");
        education.score
    }

    #[test]
    fn score_accepts_string_float_and_integer() {
        assert_eq!(score_of(r#"{"score":"3.7"}"#).as_deref(), Some("3.7"));
        assert_eq!(score_of(r#"{"score":3.7}"#).as_deref(), Some("3.7"));
        assert_eq!(score_of(r#"{"score":4}"#).as_deref(), Some("4"));
        assert_eq!(
            score_of(r#"{"score":"First Class"}"#).as_deref(),
            Some("First Class")
        );
    }

    #[test]
    fn score_missing_or_null_is_none() {
        assert_eq!(score_of("{}"), None);
        assert_eq!(score_of(r#"{"score":null}"#), None);
    }

    #[test]
    fn score_rejects_other_shapes() {
        assert!(serde_json::from_str::<Education>(r#"{"score":[3.7]}"#).is_err());
        assert!(serde_json::from_str::<Education>(r#"{"score":true}"#).is_err());
    }
}

File: shared/resume.ml
(** A JSON Resume document (jsonresume/resume-schema). Unknown properties are accepted and ignored;
    the original JSON stays the stored form. *)

open Decode

let ( let* ) = Result.bind

module Location = struct
  type t = {
    address : string option;
    postal_code : string option;
    city : string option;
    country_code : string option;
    region : string option;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* address = field "address" string ~path o in
        let* postal_code = field "postalCode" string ~path o in
        let* city = field "city" string ~path o in
        let* country_code = field "countryCode" string ~path o in
        let* region = field "region" string ~path o in
        Ok { address; postal_code; city; country_code; region })
end

module Profile = struct
  type t = { network : string option; username : string option; url : string option }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* network = field "network" string ~path o in
        let* username = field "username" string ~path o in
        let* url = field "url" string ~path o in
        Ok { network; username; url })
end

module Basics = struct
  type t = {
    name : string option;
    label : string option;
    image : string option;
    email : string option;
    phone : string option;
    url : string option;
    summary : string option;
    location : Location.t option;
    profiles : Profile.t list;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* label = field "label" string ~path o in
        let* image = field "image" string ~path o in
        let* email = field "email" string ~path o in
        let* phone = field "phone" string ~path o in
        let* url = field "url" string ~path o in
        let* summary = field "summary" string ~path o in
        let* location = field "location" Location.decode ~path o in
        let* profiles = field_list "profiles" Profile.decode ~path o in
        Ok { name; label; image; email; phone; url; summary; location; profiles })
end

module Work = struct
  type t = {
    name : string option;
    location : string option;
    description : string option;
    position : string option;
    url : string option;
    start_date : string option;
    end_date : string option;
    summary : string option;
    highlights : string list;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* location = field "location" string ~path o in
        let* description = field "description" string ~path o in
        let* position = field "position" string ~path o in
        let* url = field "url" string ~path o in
        let* start_date = field "startDate" string ~path o in
        let* end_date = field "endDate" string ~path o in
        let* summary = field "summary" string ~path o in
        let* highlights = field_list "highlights" string ~path o in
        Ok { name; location; description; position; url; start_date; end_date; summary; highlights })
end

module Volunteer = struct
  type t = {
    organization : string option;
    position : string option;
    url : string option;
    start_date : string option;
    end_date : string option;
    summary : string option;
    highlights : string list;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* organization = field "organization" string ~path o in
        let* position = field "position" string ~path o in
        let* url = field "url" string ~path o in
        let* start_date = field "startDate" string ~path o in
        let* end_date = field "endDate" string ~path o in
        let* summary = field "summary" string ~path o in
        let* highlights = field_list "highlights" string ~path o in
        Ok { organization; position; url; start_date; end_date; summary; highlights })
end

module Education = struct
  type t = {
    institution : string option;
    url : string option;
    area : string option;
    study_type : string option;
    start_date : string option;
    end_date : string option;
    score : string option;
    courses : string list;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* institution = field "institution" string ~path o in
        let* url = field "url" string ~path o in
        let* area = field "area" string ~path o in
        let* study_type = field "studyType" string ~path o in
        let* start_date = field "startDate" string ~path o in
        let* end_date = field "endDate" string ~path o in
        let* score = field "score" string_or_number ~path o in
        let* courses = field_list "courses" string ~path o in
        Ok { institution; url; area; study_type; start_date; end_date; score; courses })
end

module Award = struct
  type t = {
    title : string option;
    date : string option;
    awarder : string option;
    summary : string option;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* title = field "title" string ~path o in
        let* date = field "date" string ~path o in
        let* awarder = field "awarder" string ~path o in
        let* summary = field "summary" string ~path o in
        Ok { title; date; awarder; summary })
end

module Certificate = struct
  type t = {
    name : string option;
    date : string option;
    url : string option;
    issuer : string option;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* date = field "date" string ~path o in
        let* url = field "url" string ~path o in
        let* issuer = field "issuer" string ~path o in
        Ok { name; date; url; issuer })
end

module Publication = struct
  type t = {
    name : string option;
    publisher : string option;
    release_date : string option;
    url : string option;
    summary : string option;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* publisher = field "publisher" string ~path o in
        let* release_date = field "releaseDate" string ~path o in
        let* url = field "url" string ~path o in
        let* summary = field "summary" string ~path o in
        Ok { name; publisher; release_date; url; summary })
end

module Skill = struct
  type t = { name : string option; level : string option; keywords : string list }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* level = field "level" string ~path o in
        let* keywords = field_list "keywords" string ~path o in
        Ok { name; level; keywords })
end

module Language = struct
  type t = { language : string option; fluency : string option }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* language = field "language" string ~path o in
        let* fluency = field "fluency" string ~path o in
        Ok { language; fluency })
end

module Interest = struct
  type t = { name : string option; keywords : string list }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* keywords = field_list "keywords" string ~path o in
        Ok { name; keywords })
end

module Reference = struct
  type t = { name : string option; reference : string option }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* reference = field "reference" string ~path o in
        Ok { name; reference })
end

module Project = struct
  type t = {
    name : string option;
    description : string option;
    highlights : string list;
    keywords : string list;
    start_date : string option;
    end_date : string option;
    url : string option;
    roles : string list;
    entity : string option;
    type_ : string option;
  }

  let decode : t Decode.t =
    obj (fun ~path o ->
        let* name = field "name" string ~path o in
        let* description = field "description" string ~path o in
        let* highlights = field_list "highlights" string ~path o in
        let* keywords = field_list "keywords" string ~path o in
        let* start_date = field "startDate" string ~path o in
        let* end_date = field "endDate" string ~path o in
        let* url = field "url" string ~path o in
        let* roles = field_list "roles" string ~path o in
        let* entity = field "entity" string ~path o in
        let* type_ = field "type" string ~path o in
        Ok
          {
            name;
            description;
            highlights;
            keywords;
            start_date;
            end_date;
            url;
            roles;
            entity;
            type_;
          })
end

type t = {
  basics : Basics.t option;
  work : Work.t list;
  volunteer : Volunteer.t list;
  education : Education.t list;
  awards : Award.t list;
  certificates : Certificate.t list;
  publications : Publication.t list;
  skills : Skill.t list;
  languages : Language.t list;
  interests : Interest.t list;
  references : Reference.t list;
  projects : Project.t list;
}

let decode : t Decode.t =
  obj (fun ~path o ->
      let* basics = field "basics" Basics.decode ~path o in
      let* work = field_list "work" Work.decode ~path o in
      let* volunteer = field_list "volunteer" Volunteer.decode ~path o in
      let* education = field_list "education" Education.decode ~path o in
      let* awards = field_list "awards" Award.decode ~path o in
      let* certificates = field_list "certificates" Certificate.decode ~path o in
      let* publications = field_list "publications" Publication.decode ~path o in
      let* skills = field_list "skills" Skill.decode ~path o in
      let* languages = field_list "languages" Language.decode ~path o in
      let* interests = field_list "interests" Interest.decode ~path o in
      let* references = field_list "references" Reference.decode ~path o in
      let* projects = field_list "projects" Project.decode ~path o in
      Ok
        {
          basics;
          work;
          volunteer;
          education;
          awards;
          certificates;
          publications;
          skills;
          languages;
          interests;
          references;
          projects;
        })

let of_json json = decode ~path:"" json

(** Parse JSON text and decode it. The error carries a JSON path when the shape is wrong and an
    empty path when the text is not JSON at all. *)
let of_string text =
  let* json = Decode.parse_json text in
  of_json json

(** Section sizes for the Studio's summary tags, in contract order. *)
let section_counts t =
  List.filter
    (fun (_, n) -> n > 0)
    [
      ("work", List.length t.work);
      ("volunteer", List.length t.volunteer);
      ("education", List.length t.education);
      ("awards", List.length t.awards);
      ("certificates", List.length t.certificates);
      ("publications", List.length t.publications);
      ("skills", List.length t.skills);
      ("languages", List.length t.languages);
      ("interests", List.length t.interests);
      ("references", List.length t.references);
      ("projects", List.length t.projects);
    ]
